File: app/static/sectionlist.js
/* Section-list controller (LLM-COP-5) — the /muokkaa/osiot screen.
 *
 * The inline row editor is the same form component the /muokkaa panel
 * uses: window.createSectionForm (section-form.js), given the row's
 * .section-form container. This file forks the container, never the form.
 * The debounce is the shared one too (autosave.js), one per open row, so
 * the delay is never restated here.
 *
 * The builder mutates the draft it is handed in place and renders no
 * errors; both are the host's, as they are in edit.js. This file keeps a
 * draft and a lastSaved per open row and draws .form-errors into that
 * row's own form — never document-wide, because six rows can be open.
 *
 * Nothing here renders a summary, a badge or a section's markup. After
 * every mutation the row is re-fetched from GET /muokkaa/osiot/rivi/<id>
 * and its .row-head and .esikatselu-card are swapped in, so the summary
 * strings stay in app/summary.py and the section markup in page.html —
 * one renderer each, and this file knows neither. Three server-derived
 * things fall outside those two subtrees and are carried across rather
 * than computed here: .exp-restore's disabled flag, copied off the fresh
 * fragment; the topbar's count, taken as a finished string off the add
 * response so the Finnish plural rule stays in app/sectionlist.py; and
 * the add control's reason, rendered hidden by the template and only
 * unhidden here. Every one of them is still the server's answer.
 *
 * The row buttons are wired by one delegated click listener on the list
 * rather than per row: swapping .row-head out from under per-row
 * listeners would quietly unwire every refreshed row.
 */
(function () {
  "use strict";

  var bootstrap = JSON.parse(
    document.getElementById("bootstrap").textContent
  );
  var FIELDS = bootstrap.fields;
  var LABELS = bootstrap.field_labels;

  var list = document.querySelector(".sections-list");
  var jarjesta = document.querySelector(".jarjesta-link");
  var julkaiseButton = document.querySelector(".julkaise-button");
  var addButton = document.querySelector(".add-section-button");
  var addMenu = document.querySelector(".add-section-menu");
  var addReason = document.querySelector(".add-section-reason");
  var topbarPage = document.querySelector(".topbar-page");

  // The payload a row's form starts from, by section id. The server
  // sends it with the page and again with a newly added section, so no
  // blank payload is ever built here — that shape is the schema's.
  var payloads = {};
  bootstrap.rows.forEach(function (row) {
    payloads[row.id] = row.payload;
  });

  // Open rows only: id -> { row, draft, lastSaved, autosave, saveInFlight }
  var editors = {};

  var JSON_HEADERS = {
    "Content-Type": "application/json",
    "Accept": "application/json"
  };

  function deepCopy(value) {
    return JSON.parse(JSON.stringify(value));
  }

  function labelFor(kind, name) {
    return (LABELS[kind] || {})[name];
  }

  /* ---- rows ---- */

  function rowOf(element) {
    return element.closest(".section-row");
  }

  function rowById(id) {
    return list.querySelector('.section-row[data-section-id="' + id + '"]');
  }

  function idOf(row) {
    return Number(row.dataset.sectionId);
  }

  function allRows() {
    return Array.prototype.slice.call(
      list.querySelectorAll(".section-row")
    );
  }

  function parseRow(html) {
    // A <li> only parses inside a list element.
    var host = document.createElement("ul");
    host.innerHTML = html;
    return host.querySelector(".section-row");
  }

  // .row-head is swapped whole, so the badge, the summary and the ⋯
  // menu's can_show state all arrive fresh with it. .exp-restore does
  // not: it sits in .expanded-editor, which is deliberately left alone
  // because the open form lives there. So its one server-derived
  // attribute is carried over on its own. It has to be: POST /api/publish
  // is what first fills previous_published, and without this the button
  // an open row shows stays disabled — the one-step restore unreachable
  // in the very flow that needs it — until the page is reloaded.
  function carryRestoreState(row, fresh) {
    var button = row.querySelector(".exp-restore");
    var source = fresh.querySelector(".exp-restore");
    // The server decides can_restore; this reads its answer off the
    // fragment rather than working it out a second time here.
    if (button && source) button.disabled = source.disabled;
  }

  function refreshRow(id) {
    return fetch("/muokkaa/osiot/rivi/" + id, {
      headers: { "Accept": "text/html" }
    }).then(function (response) {
      if (!response.ok) return;
      return response.text().then(function (html) {
        var fresh = parseRow(html);
        var row = rowById(id);
        if (!fresh || !row) return;
        // Only these two subtrees, never the whole <li>: an open form and
        // the caret inside it have to survive a save.
        row
          .querySelector(".row-head")
          .replaceWith(fresh.querySelector(".row-head"));
        row
          .querySelector(".esikatselu-card")
          .replaceWith(fresh.querySelector(".esikatselu-card"));
        carryRestoreState(row, fresh);
      });
    });
  }

  function refreshAllRows() {
    return Promise.all(
      allRows().map(function (row) {
        return refreshRow(idOf(row));
      })
    );
  }

  /* ---- the inline editor ---- */

  function formOf(editor) {
    return editor.row.querySelector(".section-form");
  }

  // Error rendering is the host's; the shared builder draws controls and
  // nothing else. The box is scoped to the row's own form, because every
  // row can be open at once and a document-wide lookup would file one
  // row's refusal under another row's fields.
  function showErrors(editor, errors) {
    var form = formOf(editor);
    var box = form.querySelector(".form-errors");
    if (!box) {
      box = document.createElement("div");
      box.className = "form-errors";
      form.insertBefore(box, form.firstChild);
    }
    var kind = editor.row.dataset.kind;
    box.textContent = Object.keys(errors)
      .map(function (name) {
        return (labelFor(kind, name) || name) + ": " + errors[name];
      })
      .join(" · ");
  }

  function clearErrors(editor) {
    var box = formOf(editor).querySelector(".form-errors");
    if (box) box.remove();
  }

  // Redrawing is how a payload is replaced: the builder empties the mount
  // and binds every control to the draft it is handed, so Peruuta and
  // Palauta swap editor.draft and call this rather than setting values.
  function buildForm(editor) {
    window.createSectionForm(formOf(editor), {
      kind: editor.row.dataset.kind,
      draft: editor.draft,
      fields: FIELDS,
      labels: LABELS,
      onChange: function () {
        editor.autosave.schedule();
      }
    });
  }

  function openEditor(row) {
    var id = idOf(row);
    if (!editors[id]) {
      var editor = {
        row: row,
        // The builder mutates the draft in place, so the bootstrapped
        // payload is copied rather than handed over to be typed into.
        draft: deepCopy(payloads[id]),
        lastSaved: deepCopy(payloads[id]),
        autosave: null,
        saveInFlight: null
      };
      // save is a hoisted declaration, so it is safe to name here.
      editor.autosave = window.createAutosave(
        window.createAutosave.DELAY,
        function () {
          save(id);
        }
      );
      editors[id] = editor;
      buildForm(editor);
    }
    row.querySelector(".expanded-editor").hidden = false;
  }

  function closeEditor(row) {
    var editor = editors[idOf(row)];
    // A debounced change has to land before the row folds away: the fresh
    // .row-head that save fetches carries the summary the collapsed row
    // shows, so closing on a pending write would show a stale one.
    if (editor) editor.autosave.flush();
    row.querySelector(".expanded-editor").hidden = true;
  }

  function showSavedTime(editor, savedAt) {
    var when = new Date(savedAt * 1000);
    editor.row.querySelector(".exp-autosave-time").textContent =
      when.getHours() + "." + ("0" + when.getMinutes()).slice(-2);
  }

  function save(id) {
    var editor = editors[id];
    // A save now settles whatever the debounce had queued.
    editor.autosave.cancel();
    // The existing LLM-COP-4 draft route, so this write goes through the
    // same validator and the same serialization convention as the panel's.
    var payload = deepCopy(editor.draft);
    var request = fetch("/api/sections/" + id + "/draft", {
      method: "PUT",
      headers: JSON_HEADERS,
      body: JSON.stringify(payload)
    }).then(function (response) {
      if (response.status === 401) {
        window.location = "/yllapito";
        return;
      }
      return response.json().then(function (data) {
        if (!response.ok) {
          showErrors(editor, data.errors || {});
          return;
        }
        clearErrors(editor);
        editor.lastSaved = payload;
        showSavedTime(editor, data.saved_at);
        return refreshRow(id);
      });
    });
    editor.saveInFlight = request;
    request.then(
      function () {
        if (editor.saveInFlight === request) editor.saveInFlight = null;
      },
      function () {
        if (editor.saveInFlight === request) editor.saveInFlight = null;
      }
    );
    return request;
  }

  function cancel(id) {
    var editor = editors[id];
    if (!editor) return;
    editor.autosave.cancel();
    editor.draft = deepCopy(editor.lastSaved);
    clearErrors(editor);
    buildForm(editor);
  }

  function restore(id) {
    fetch("/api/sections/" + id + "/restore", {
      method: "POST",
      headers: JSON_HEADERS
    }).then(function (response) {
      if (response.status === 401) {
        window.location = "/yllapito";
        return;
      }
      return response.json().then(function (data) {
        if (!response.ok) return;
        // The restore landed in the draft only; the row's badge and card
        // come back from the fragment, and the disclaimer beside them
        // already says a Julkaise is still needed.
        var editor = editors[id];
        if (editor) {
          editor.autosave.cancel();
          editor.draft = deepCopy(data.payload);
          editor.lastSaved = deepCopy(data.payload);
          clearErrors(editor);
          buildForm(editor);
        }
        payloads[id] = data.payload;
        return refreshRow(id);
      });
    });
  }

  function setState(id, state) {
    fetch("/api/sections/" + id + "/state", {
      method: "POST",
      headers: JSON_HEADERS,
      body: JSON.stringify({ state: state })
    }).then(function (response) {
      if (response.status === 401) {
        window.location = "/yllapito";
        return;
      }
      // A refusal is rendered by the row itself — the fresh fragment
      // carries the badge and the menu's disabled reason either way.
      return refreshRow(id);
    });
  }

  function toggleMenu(row) {
    var items = row.querySelector(".row-menu-items");
    var opening = items.hidden;
    list.querySelectorAll(".row-menu-items").forEach(function (other) {
      other.hidden = true;
    });
    items.hidden = !opening;
  }

  /* ---- one delegated listener for every row control ---- */

  list.addEventListener("click", function (event) {
    var trigger = event.target.closest("[data-action]");
    if (!trigger || !list.contains(trigger)) return;
    var row = rowOf(trigger);
    if (!row) return;
    var id = idOf(row);
    var action = trigger.dataset.action;
    if (action === "edit") {
      openEditor(row);
    } else if (action === "close") {
      closeEditor(row);
    } else if (action === "menu") {
      toggleMenu(row);
    } else if (action === "hide") {
      setState(id, "hidden");
    } else if (action === "show") {
      setState(id, "published");
    } else if (action === "save") {
      if (editors[id]) save(id);
    } else if (action === "cancel") {
      cancel(id);
    } else if (action === "restore") {
      restore(id);
    }
  });

  /* ---- reordering ---- */

  function applyOrder() {
    // The whole order, read off the DOM — the endpoint refuses anything
    // shorter, because a partial list is a row whose position nobody set.
    var ids = allRows().map(idOf);
    return fetch("/api/sections/order", {
      method: "PUT",
      headers: JSON_HEADERS,
      body: JSON.stringify({ ids: ids })
    }).then(function (response) {
      if (response.status === 401) window.location = "/yllapito";
    });
  }

  function reordering() {
    return list.classList.contains("reordering");
  }

  // draggable is the whole of what makes a real mouse drag possible, and
  // every row's flag is derived from the list's mode here and nowhere
  // else. Re-applied rather than toggled, so a row added while Järjestä
  // osiot is on is draggable at once instead of only after the mode is
  // switched off and on again.
  function applyDraggable() {
    var on = reordering();
    allRows().forEach(function (row) {
      row.draggable = on;
    });
  }

  jarjesta.addEventListener("click", function (event) {
    event.preventDefault();
    list.classList.toggle("reordering");
    applyDraggable();
  });

  list.addEventListener("dragstart", function (event) {
    var row = rowOf(event.target);
    if (!row || !reordering()) return;
    // dataTransfer, not a module variable: the id travels in the drag
    // itself, so nothing depends on setDragImage or effectAllowed.
    event.dataTransfer.setData("text/plain", String(idOf(row)));
  });

  list.addEventListener("dragover", function (event) {
    if (!reordering()) return;
    event.preventDefault(); // without this no drop ever fires
  });

  list.addEventListener("drop", function (event) {
    if (!reordering()) return;
    event.preventDefault();
    var target = rowOf(event.target);
    var dragged = rowById(Number(event.dataTransfer.getData("text/plain")));
    if (!target || !dragged || target === dragged) return;
    var rows = allRows();
    if (rows.indexOf(dragged) < rows.indexOf(target)) {
      target.after(dragged);
    } else {
      target.before(dragged);
    }
    applyOrder();
  });

  list.addEventListener("keydown", function (event) {
    if (event.key !== "ArrowUp" && event.key !== "ArrowDown") return;
    var handle = event.target.closest(".row-drag-handle");
    if (!handle) return;
    var row = rowOf(handle);
    if (event.key === "ArrowUp" && row.previousElementSibling) {
      row.previousElementSibling.before(row);
    } else if (event.key === "ArrowDown" && row.nextElementSibling) {
      row.nextElementSibling.after(row);
    } else {
      return;
    }
    event.preventDefault();
    handle.focus();
    applyOrder();
  });

  /* ---- adding a section ---- */

  // Taking the last offerable kind closes the control: the button goes
  // disabled and the reason beside it explains why. The sentence is not
  // written here — edit_sections.html renders it on every page and hides it
  // while a kind is still on offer, so this only flips the attribute, the
  // way carryRestoreState only copies one.
  //
  // Both functions here dereference addButton and addMenu, which are null
  // on a seeded install where no kind is offerable. Neither is ever called
  // in that case: the only caller is the listener below, wired inside the
  // same guard that establishes both.
  function closeAddIfEmpty() {
    if (addMenu.querySelector(".add-section-item")) return;
    addButton.disabled = true;
    addMenu.hidden = true;
    if (addReason) addReason.hidden = false;
  }

  function addSection(item) {
    addMenu.hidden = true;
    return fetch("/api/sections", {
      method: "POST",
      headers: JSON_HEADERS,
      body: JSON.stringify({ kind: item.dataset.kind })
    }).then(function (response) {
      if (response.status === 401) {
        window.location = "/yllapito";
        return;
      }
      return response.json().then(function (data) {
        if (!response.ok) return;
        payloads[data.id] = data.payload;
        list.insertAdjacentHTML("beforeend", data.html);
        applyDraggable();
        // The topbar counts the rows, so appending one made it wrong. The
        // fresh string is the server's — the plural rule stays in
        // app/sectionlist.py beside the one the page was rendered with.
        topbarPage.textContent = data.page_label;
        item.remove();
        closeAddIfEmpty();
      });
    });
  }

  if (addButton && addMenu) {
    addButton.addEventListener("click", function () {
      addMenu.hidden = !addMenu.hidden;
    });

    addMenu.addEventListener("click", function (event) {
      var item = event.target.closest(".add-section-item");
      if (!item) return;
      addSection(item);
    });
  }

  /* ---- Julkaise ---- */

  julkaiseButton.addEventListener("click", function () {
    // Every pending autosave — debounced or in flight — must land before
    // the publish reads the drafts.
    var pending = [];
    Object.keys(editors).forEach(function (id) {
      // flush() hands back save()'s own request when one was debounced,
      // and null when none was — so an in-flight PUT is waited on in its
      // place and a quiet row contributes nothing.
      var editor = editors[id];
      var landing = editor.autosave.flush() || editor.saveInFlight;
      if (landing) pending.push(landing);
    });
    Promise.all(pending).then(publishNow);
  });

  function publishNow() {
    return fetch("/api/publish", {
      method: "POST",
      headers: { "Accept": "application/json" }
    }).then(function (response) {
      if (response.status === 401) {
        window.location = "/yllapito";
        return;
      }
      return response.json().then(function () {
        return refreshAllRows();
      });
    });
  }
})();

File: app/sectionlist.py
"""Section-list mode (LLM-COP-5): the /muokkaa/osiot screen, the row
fragment route, and the reorder / add / hide-show / restore API. Every
route is admin-gated with auth.require_admin.

Two conventions this module inherits rather than invents. Drafts are
written with json.dumps(..., ensure_ascii=False) in FIELDS[kind]
declaration order (app/edit.py:4-8) — a new section's blank draft and a
restore both go through it, so the badge's and publish's text-equality
comparisons stay honest. And the badge itself is app.sections.badge; this
module computes no badge of its own.

The row fragment is why there is a route per row: the draft PUT answers
only saved_at and badge (app/edit.py:96-99), so a client that wanted a
fresh summary or a fresh preview card would build both itself — the
summary strings in JavaScript beside app/summary.py, the six sections'
markup beside app/templates/page.html. Instead the client re-fetches
GET /muokkaa/osiot/rivi/<id> after every mutation and swaps two subtrees
out of it, and the server stays the only place either exists.

The refusal strings below are free text, not spec addresses: no criterion
in cp-main-edit-sections or cp-section-row asserts any of them.
"""

import json
import time

from flask import Blueprint, current_app, jsonify, render_template, request

from . import auth
from . import db as database
from .fields import ANCHORS, FIELD_LABELS, FIELDS, SECTION_NAMES
from .images import collect_unreferenced
from .sanitize import validate_payload
from .sections import badge
from .summary import blank_payload, summarize

bp = Blueprint("sectionlist", __name__)

# Näytä osio is refused when the row has nothing publishable to show.
# Each refusal carries the reason and the one thing that clears it, and
# the row menu renders the hint as its visible disabled reason.
NO_PUBLISHED = "osiolla ei ole julkaistua sisältöä"
NO_PUBLISHED_HINT = "Julkaise osio ensin"
BLANK_PUBLISHED = "osion sisältö on tyhjä"
BLANK_PUBLISHED_HINT = "Lisää sisältöä ja julkaise"

NO_PREVIOUS = "osiolla ei ole edellistä julkaistua versiota"
NO_PREVIOUS_HINT = "Julkaise osio ensin"

# PUT /api/sections/order takes the whole order or nothing: a partial
# list is a lost row, not a smaller write.
ORDER_MALFORMED = "ids puuttuu tai ei ole lista kokonaislukuja"
ORDER_NOT_WHOLE = "listassa on oltava jokainen osio täsmälleen kerran"

UNKNOWN_KIND = "tuntematon osiotyyppi"
KIND_PRESENT = "osiotyyppi on jo sivulla"
UNKNOWN_STATE = "tuntematon tila"
NOT_FOUND = "not found"


def _connect():
    return database.connect(current_app.config["DATABASE"])


def page_label(count):
    """"Etusivu · 6 osiota" — the topbar's count with its plural.

    Here rather than inline in edit_sections.html because adding a section
    changes it without a reload: the add response carries the fresh string
    and the client writes it in, exactly as it carries the badge and the
    summary. A client that counted its own rows would have to carry the
    Finnish plural rule too, and the two would drift.
    """
    return f"Etusivu · {count} {'osio' if count == 1 else 'osiota'}"


def _show_refusal(kind, published):
    """Why this row may not be flipped to state='published', or None.

    The public page gates on state alone and falls back to an empty
    payload (app/sections.py:22-24 and :30), so a row whose published
    column is NULL or blank renders as literally nothing once it is shown.
    The gate is on the transition, and it is this one expression — used
    both to answer the flip and to draw the menu item that requests it.
    """
    if published is None:
        return NO_PUBLISHED, NO_PUBLISHED_HINT
    if json.loads(published) == blank_payload(kind):
        return BLANK_PUBLISHED, BLANK_PUBLISHED_HINT
    return None


def list_rows(conn):
    """Every section in page order, with all a row needs to render itself.

    Its own SELECT rather than app.sections.draft_sections, because a row
    needs two flags the edit panel never wanted — can_restore and can_show
    — and app/sections.py is a shared file the units landing beside this
    one are also editing. The duplication is knowing, and named here.
    """
    rows = conn.execute(
        "SELECT id, kind, position, state, draft, published,"
        " previous_published FROM sections ORDER BY position"
    ).fetchall()
    out = []
    for row in rows:
        # The third copy of this fallback (app/sections.py:56 is the
        # sibling to keep it in step with): a row whose draft was never
        # written reads as the empty payload rather than raising.
        payload = json.loads(row["draft"]) if row["draft"] else {}
        refusal = _show_refusal(row["kind"], row["published"])
        out.append(
            {
                "id": row["id"],
                "kind": row["kind"],
                "position": row["position"],
                "state": row["state"],
                "badge": badge(row["state"], row["draft"], row["published"]),
                "payload": payload,
                "summary": summarize(row["kind"], payload),
                "can_restore": row["previous_published"] is not None,
                "can_show": refusal is None,
                "show_reason": None if refusal is None else refusal[1],
            }
        )
    return out


def _one_row(conn, section_id):
    """One row in exactly the shape list_rows gives it.

    Deliberately the same builder rather than a second SELECT: the row
    fragment and the row inside the page document must be byte-identical,
    and the cheapest way to guarantee that is one code path.
    """
    for row in list_rows(conn):
        if row["id"] == section_id:
            return row
    return None


def _render_row(section):
    return render_template(
        "_row_fragment.html",
        section=section,
        section_names=SECTION_NAMES,
        anchors=ANCHORS,
    )


@bp.route("/muokkaa/osiot")
@auth.require_admin
def osiot():
    """The section list: every row server-rendered in position order.

    The context is deliberately narrow. _section_row.html imports
    page.html to reuse the public macros, and an import with context
    executes that template's module body — which iterates `sections` and
    dereferences `nav_labels` while doing it. Neither name is bound here
    (the row list is `rows`), so the module body's loops find an undefined
    iterable, yield nothing, and their output is discarded. `anchors` is
    bound because every public macro dereferences it.
    """
    conn = _connect()
    try:
        rows = list_rows(conn)
    finally:
        conn.close()
    bootstrap = {
        "rows": rows,
        "fields": FIELDS,
        "field_labels": FIELD_LABELS,
        "section_names": SECTION_NAMES,
        "anchors": ANCHORS,
    }
    return render_template(
        "edit_sections.html",
        rows=rows,
        page_label=page_label(len(rows)),
        section_names=SECTION_NAMES,
        anchors=ANCHORS,
        bootstrap=bootstrap,
    )


@bp.route("/muokkaa/osiot/rivi/<int:section_id>")
@auth.require_admin
def rivi(section_id):
    """One row's HTML, re-fetched by the client after every mutation."""
    conn = _connect()
    try:
        section = _one_row(conn, section_id)
    finally:
        conn.close()
    if section is None:
        return jsonify(error=NOT_FOUND), 404
    return _render_row(section)


def _order_error(conn, body):
    """Why this reorder is refused, or None — checked before any write.

    The whole order arrives or nothing does: a list that is missing a row,
    names one twice, or names a row that does not exist would silently
    leave positions behind, and a position nobody set is a page order
    nobody chose.
    """
    if not isinstance(body, dict):
        return ORDER_MALFORMED
    ids = body.get("ids")
    if not isinstance(ids, list):
        return ORDER_MALFORMED
    for section_id in ids:
        # bool is an int in Python; it is not a section id.
        if not isinstance(section_id, int) or isinstance(section_id, bool):
            return ORDER_MALFORMED
    if len(set(ids)) != len(ids):
        return ORDER_NOT_WHOLE
    known = {
        row["id"] for row in conn.execute("SELECT id FROM sections").fetchall()
    }
    if set(ids) != known:
        return ORDER_NOT_WHOLE
    return None


@bp.route("/api/sections/order", methods=["PUT"])
@auth.require_admin
def put_order():
    """Whole-order write: positions 1..n follow the ids as sent.

    sqlite3 opens the implicit transaction at the first UPDATE and holds
    it to the single commit below, so every position lands or none does.
    auth.audit runs after that commit, never between the updates, because
    it commits internally (app/auth.py:103-113) — the ordering
    app/edit.py:94-95 already uses.
    """
    body = request.get_json(silent=True)
    conn = _connect()
    try:
        error = _order_error(conn, body)
        if error is not None:
            return jsonify(error=error), 400
        ids = body["ids"]
        for position, section_id in enumerate(ids, start=1):
            conn.execute(
                "UPDATE sections SET position = ? WHERE id = ?",
                (position, section_id),
            )
        conn.commit()
        auth.audit(
            conn, "sections reordered ids=" + ",".join(str(i) for i in ids)
        )
        return jsonify(ids=ids)
    finally:
        conn.close()


@bp.route("/api/sections", methods=["POST"])
@auth.require_admin
def post_section():
    """Add one section of a kind the page does not have yet.

    It arrives hidden, last in the order, with a blank draft and nothing
    published — badge Piilotettu — so adding a section and then never
    publishing it leaves the public page exactly as it was.
    """
    body = request.get_json(silent=True) or {}
    kind = body.get("kind")
    conn = _connect()
    try:
        if kind not in FIELDS:
            return jsonify(error=UNKNOWN_KIND), 400
        present = {
            row["kind"]
            for row in conn.execute("SELECT kind FROM sections").fetchall()
        }
        if kind in present:
            return jsonify(error=KIND_PRESENT), 400
        # Through the same validator the draft PUT uses, so the row's
        # first stored text is exactly what a no-op save would produce.
        clean, errors = validate_payload(kind, blank_payload(kind))
        if errors:  # pragma: no cover - the schema's own zero values
            return jsonify(errors=errors), 400
        (top,) = conn.execute(
            "SELECT COALESCE(MAX(position), 0) FROM sections"
        ).fetchone()
        cursor = conn.execute(
            "INSERT INTO sections (kind, position, state, draft, published,"
            " previous_published) VALUES (?, ?, 'hidden', ?, NULL, NULL)",
            (kind, top + 1, json.dumps(clean, ensure_ascii=False)),
        )
        conn.commit()
        section_id = cursor.lastrowid
        auth.audit(conn, f"section added kind={kind} id={section_id}")
        rows = list_rows(conn)
        section = next(row for row in rows if row["id"] == section_id)
        count = len(rows)
    finally:
        conn.close()
    return (
        jsonify(
            id=section["id"],
            kind=section["kind"],
            state=section["state"],
            badge=section["badge"],
            # The blank payload the row's form starts from, so the client
            # never has to build one and keep a copy of the schema.
            payload=section["payload"],
            # The topbar's count, freshly rendered: the row the client is
            # about to append makes the served one wrong, and this is the
            # same string the reloaded screen would show.
            page_label=page_label(count),
            html=_render_row(section),
        ),
        201,
    )


@bp.route("/api/sections/<int:section_id>/state", methods=["POST"])
@auth.require_admin
def post_state(section_id):
    """Hide or show one section — state only, nothing else is touched."""
    body = request.get_json(silent=True) or {}
    state = body.get("state")
    if state not in ("hidden", "published"):
        return jsonify(error=UNKNOWN_STATE), 400
    conn = _connect()
    try:
        row = conn.execute(
            "SELECT id, kind, draft, published FROM sections WHERE id = ?",
            (section_id,),
        ).fetchone()
        if row is None:
            return jsonify(error=NOT_FOUND), 404
        if state == "published":
            refusal = _show_refusal(row["kind"], row["published"])
            if refusal is not None:
                return jsonify(error=refusal[0], hint=refusal[1]), 409
        conn.execute(
            "UPDATE sections SET state = ? WHERE id = ?", (state, section_id)
        )
        conn.commit()
        auth.audit(conn, f"section state {state} id={section_id}")
        return jsonify(badge=badge(state, row["draft"], row["published"]))
    finally:
        conn.close()


@bp.route("/api/sections/<int:section_id>/restore", methods=["POST"])
@auth.require_admin
def post_restore(section_id):
    """Palauta edellinen versio: previous_published back into the DRAFT.

    One step, and only that step — published, previous_published and
    state are left alone, so the restore still needs a Julkaise to reach
    the public page and the screen says so. The bytes are copied
    verbatim rather than re-validated and re-serialized: they were
    produced by the same json.dumps convention, so a restore followed by
    Julkaise puts back a published payload byte-identical to the one that
    was there before.
    """
    conn = _connect()
    try:
        row = conn.execute(
            "SELECT id, state, published, previous_published FROM sections"
            " WHERE id = ?",
            (section_id,),
        ).fetchone()
        if row is None:
            return jsonify(error=NOT_FOUND), 404
        if row["previous_published"] is None:
            return jsonify(error=NO_PREVIOUS, hint=NO_PREVIOUS_HINT), 409
        text = row["previous_published"]
        conn.execute(
            "UPDATE sections SET draft = ? WHERE id = ?", (text, section_id)
        )
        conn.commit()
        auth.audit(conn, f"section restored id={section_id}")
        # The replaced draft's digest may have been this store's last
        # reference to a picture (LLM-COP-27).
        collect_unreferenced(conn)
        return jsonify(
            badge=badge(row["state"], text, row["published"]),
            restored_at=int(time.time()),
            payload=json.loads(text),
        )
    finally:
        conn.close()
